webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/nodes/MergeNodeProperties.scala

```diff
@@ -349,87 +314,94 @@ object MergeNodeProperties {
+  /**
+   *
+   * Sort groups.
+   * 1/ Start by finding all graph (almost tree) of groups. That part can fail if any parent
+   * referenced in a group criterium is missing in the parameter list.
+   *
+   * Edges are defined thanks to two properties:
+   * - if S is a subgroup of P, then there's an edge from P to S (P -> S),
+   * - if S is subgroup of P1 (line 1) and P2 (line 2), then there's an edge from P1 to P2 (P1 -> p2)
+   *
+   * 2/ Then, for each graph, return the topological sort of its GroupProp. That part can
+   * fail is there is cycle in any graph.
+   *
+   * The sort returns the oldest parent first.
    *
-   * When we are at root, we need to go back up, actually building the tree with all
-   * its branches and not only the one we took during descent:
-   *   - look at each line in criterion (we only have subgroup criterion here), keeping line
-   *     order (because it defines override order):
-   *     - 0 means we have the outer most parent
-   *     - n > 0 means: more parents!
    */
-  def buildOneTree(group: GroupProp, allTargets: Map[NodeGroupId, GroupProp]): PureResult[InvertedTree[GroupProp]] = {
-    /*
-     * Descend to root
-     */
-    @scala.annotation.tailrec
-    def searchDown(current: GroupProp, all: Map[NodeGroupId, GroupProp]): PureResult[InvertedTree[GroupProp]] = {
-      all.collect { case g if g._2.parentGroups.contains(current.groupId.value) => g._2 } match {
-        case Nil      => // found root, go up now
-          buildUp(current, all)
-        case h :: _ => // recurse
-          searchDown(h, all)
-      }
-    }
-
-    /*
-     * Here, we look up for parent groups. We know that "all" contains all parents.
-     */
-    def buildUp(root: GroupProp, all: Map[NodeGroupId, GroupProp]): PureResult[InvertedTree[GroupProp]] = {
-      // recursively takes parent, keeping criterion line order
+  def sortGroups(groups: Map[String, GroupProp]): PureResult[List[List[GroupProp]]] = {
+    type GRAPH = DefaultDirectedGraph[GroupProp, DefaultEdge]
+    // Recursively add groups in the list. The last group is the oldest parent.
+    // `addEdge` can throw exception if one of the vertice is missing (it can only be the parent in our case)
+    def recAddEdges(graph: GRAPH, child: GroupProp, parents: List[GroupProp]): PureResult[GRAPH] = {
       for {
-        parents <- root.parentGroups.accumulatePure { id =>
-                     all.get(NodeGroupId(id)) match {
-                       case None    => Left(Inconsistency(s"Group '${root.groupName}' (${root.groupId.value}) has a group comparator " +
-                                                       s"criterion but the corresponding group was not found: ${id}"))
-                       case Some(g) => buildUp(g, all)
-                     }
-                   }
-      } yield {
-        InvertedTree(root, parents)
-      }
+        _ <- parents.traverse { p =>
+               // P -> S
+               try {
```

Comment:
You could use 

```
IOResult.effect(
  s"Error` when trying to build group hierarchy of group '${child.groupName}'  (${child.groupId.value}) for parent '${p.groupName}' (${p.groupId.value})"
) (graph.addEdge(p,child))
```

Don't you think ?

Reply:
With a .either, i forgot it in my example

Reply:
the exception are as pure as possible :) (ie: they are just here to denote some information about the code structure, something that would be done by an either in a better language than java). So I would let things in `PureResult` domain, and not bring the whole `ZIO` runtime for that.

Reply:
Perfect :)

Reply:
I'm ok with the "not brin the whole zio stack within for this" ;) And pure Exception because not IO... That Should be an execption Type PureException

Reply:
just to be clear - an exception is as far from purity as we can go, it breaks referential transparency, lead to horror in concurrent programming, etc.
Nonetheless, this case seems ok since: there's no concurrent context here (all the mutable parts are in the function and don't span longer than it, nothing shard, etc) ; exception only denotes an immutable property on the code and so are purelly deterministic.